test/cppsuite/tests/cursor_bound_01.cpp

```diff
@@ -0,0 +1,672 @@
+        if (next) {
+            range_ret = range_cursor->next(range_cursor.get());
+            /*
+             * If the key exists, position the cursor to the lower key using search near otherwise
+             * use prev().
+             */
+            if (lower_bound.key.length() != 0) {
+                normal_cursor->set_key(normal_cursor.get(), lower_bound.key.c_str());
+                normal_ret = normal_cursor->search_near(normal_cursor.get(), &exact);
+                if (normal_ret == WT_NOTFOUND)
+                    return;
+                if (exact < 0)
+                    normal_ret = normal_cursor->next(normal_cursor.get());
+            } else
+                normal_ret = normal_cursor->next(normal_cursor.get());
+        } else {
+            range_ret = range_cursor->prev(range_cursor.get());
+            /*
+             * If the key exists, position the cursor to the upper key using search near otherwise
+             * use next().
+             */
+            if (upper_bound.key.length() != 0) {
+                normal_cursor->set_key(normal_cursor.get(), upper_bound.key.c_str());
+                normal_ret = normal_cursor->search_near(normal_cursor.get(), &exact);
+                if (normal_ret == WT_NOTFOUND)
+                    return;
+                if (exact > 0)
+                    normal_ret = normal_cursor->prev(normal_cursor.get());
+            } else
+                normal_ret = normal_cursor->prev(normal_cursor.get());
+        }
+
+        if (normal_ret == WT_NOTFOUND)
+            return;
+
+        testutil_assert(normal_ret == range_ret && (normal_ret == 0 || normal_ret == WT_NOTFOUND));
+
+        /* Retrieve the key the cursor is pointing at. */
+        const char *normal_key, *range_key;
+        testutil_check(normal_cursor->get_key(normal_cursor.get(), &normal_key));
+        testutil_check(range_cursor->get_key(range_cursor.get(), &range_key));
+        testutil_assert(strcmp(range_key, normal_key) == 0);
+        while (true) {
+            if (next) {
+                normal_ret = normal_cursor->next(normal_cursor.get());
+                range_ret = range_cursor->next(range_cursor.get());
+            } else {
+                normal_ret = normal_cursor->prev(normal_cursor.get());
+                range_ret = range_cursor->prev(range_cursor.get());
+            }
+            testutil_assert(normal_ret == 0 || normal_ret == WT_NOTFOUND);
+            testutil_assert(range_ret == 0 || range_ret == WT_NOTFOUND);
+
+            /* Early exit if we have reached the end of the table. */
+            if (range_ret == WT_NOTFOUND && normal_ret == WT_NOTFOUND)
+                break;
+
+            /* It is possible that we have reached the end of the bounded range. */
+            else if (range_ret == WT_NOTFOUND && normal_ret == 0) {
+                testutil_check(normal_cursor->get_key(normal_cursor.get(), &normal_key));
+                std::string normal_key_str = normal_key;
```

Comment:
You don't need to create `normal_key_str`. Passing a `const char*` to `custom_lexicographical_compare` is ok to do, the conversion will be implicitly done.